client/order/epochqueue.go

```diff
@@ -0,0 +1,190 @@
+	ci := sort.Search(len(eq.commitmentIndex), func(i int) bool {
+		order := eq.orders[eq.commitmentIndex[i]]
+		return bytes.Compare(order.Commitment[:], note.Commitment[:]) > 0
+	})
+	eq.commitmentIndex = append(eq.commitmentIndex, order.OrderID{})
+	copy(eq.commitmentIndex[ci+1:], eq.commitmentIndex[ci:])
+	eq.commitmentIndex[ci] = oid
+	eq.commitmentIndexMtx.Unlock()
+}
+
+// Exists checks if the provided order id is in the queue.
+func (eq *EpochQueue) Size() int {
+	eq.ordersMtx.Lock()
+	size := len(eq.orders)
+	eq.ordersMtx.Unlock()
+	return size
+}
+
+// Exists checks if the provided order id is in the queue.
+func (eq *EpochQueue) Exists(oid order.OrderID) bool {
+	eq.ordersMtx.Lock()
+	_, ok := eq.orders[oid]
+	eq.ordersMtx.Unlock()
+	return ok
+}
+
+// GenerateMatchProof calculates the sorting seed used in order matching as well as the
+// commitment checksum from the provided epoch queue preimages and misses.
+func (eq *EpochQueue) GenerateMatchProof(preimages []order.Preimage, misses []string) (int64, msgjson.Bytes, error) {
+	eq.ordersMtx.Lock()
+	defer eq.ordersMtx.Unlock()
+
+	// Remove all misses.
+	for _, entry := range misses {
+		b, err := hex.DecodeString(entry[:])
+		if err != nil {
+			return 0, nil, fmt.Errorf("decoding error: %v", err)
+		}
+		var oid order.OrderID
+		copy(oid[:], b[:])
+
+		delete(eq.orders, oid)
+
+		// Remove the seed index associated with the miss, preserving
+		// the sort order in the process.
+		eq.seedIndexMtx.Lock()
+		si := sort.Search(len(eq.seedIndex), func(i int) bool { return bytes.Compare(eq.seedIndex[i][:], oid[:]) >= 0 })
+		if si == len(eq.seedIndex) {
+			return 0, nil, fmt.Errorf("expected to find id %s in seed sort index", oid)
+		}
+		if si < len(eq.seedIndex)-1 {
+			copy(eq.seedIndex[si:], eq.seedIndex[si+1:])
+		}
+		eq.seedIndex[len(eq.seedIndex)-1] = order.OrderID{}
+		eq.seedIndex = eq.seedIndex[:len(eq.seedIndex)-1]
+		eq.seedIndexMtx.Unlock()
+
+		// Remove the commitment index associated with the miss, preserving
+		// the sort order in the process.
+		eq.commitmentIndexMtx.Lock()
+		ci := sort.Search(len(eq.commitmentIndex), func(i int) bool { return bytes.Compare(eq.commitmentIndex[i][:], oid[:]) >= 0 })
+		if ci == len(eq.commitmentIndex) {
+			return 0, nil, fmt.Errorf("expected to find id %s in commitment sort index", oid)
+		}
+		if ci < len(eq.commitmentIndex)-1 {
+			copy(eq.commitmentIndex[ci:], eq.commitmentIndex[ci+1:])
+		}
+		eq.commitmentIndex[len(eq.commitmentIndex)-1] = order.OrderID{}
+		eq.commitmentIndex = eq.commitmentIndex[:len(eq.commitmentIndex)-1]
+		eq.commitmentIndexMtx.Unlock()
+	}
+
+	// Map the preimages received with their associated epoch order ids.
+	matches := make(map[order.OrderID]msgjson.Bytes)
+	for i := 0; i < len(preimages); i++ {
+		for oid, note := range eq.orders {
+			commitment := blake256.Sum256(preimages[i][:])
+			if bytes.Equal(note.Commitment, commitment[:]) {
+				matches[oid] = preimages[i][:]
+				break
```

Comment:
Perhaps `continue outerLoop` and then a warning after this inner loop because that would indicate an order for preimage X was not found.